add docs on query page (#7066)
fishtown-analytics/docs.getdbt.com: website/docs/docs/collaborate/access-query-page.md

---
title: "Access the Query page"
description: "Learn how to access the Query page and run queries"
sidebar_label: "Access and run queries"
tags: [Query page]
---

# Access the Query page interface <Lifecycle status="beta" />

<IntroText>
Learn how to access the Query page, run queries, and view results.
</IntroText>

:::tip
Available in private beta. To join, please reach out to your account manager. To provide feedback, use this feedback form…
:::

The Query page provides a rich console experience with editor navigation. You can expect the Query page to:

> - Have an editor to enable you to write SQL queries, with the option to open multiple tabs 

Reviewer: ```suggestion
- Enable you to write SQL queries, with the option to open multiple tabs 
```

> - View highlighted keywords and parentheses, along with basic code coloring to help with readability 

Reviewer: ```suggestion
- Have SQL + dbt autocomplete suggestions and syntax highlighting
```

> - Save or run the SQL queries you prefer

Reviewer: ```suggestion
- Bookmark SQL queries
```

- View the results of the query and its details using the **Results** or **Details** tabs

> - View the chart of the query results using the **Chart** tab

Reviewer: ```suggestion
- Create a visualization of your query results using the **Chart** tab
```

- View the history of queries and their statuses (like Success, Error, Pending) using the **Query history** icon
- Integrate with dbt Explorer, dbt Cloud IDE, and Visual Editor to provide a seamless experience for your data exploration and collaboration.

## Access the Query page

1. To access the Query page, select the **Query** option in the navigation sidebar.
2. If your [developer credentials](/docs/cloud/dbt-cloud-ide/develop-in-the-cloud#get-started-with-the-cloud-ide) aren’t set up, the Query page will prompt you to set them up. 
3. Once your credentials are set up, you can write, run, and edit SQL queries in the Query page editor for existing models in your project.

## Run queries

To run queries in the Query page, you can use:
- Standard SQL  
- Jinja (`ref`, `source`, `is_incremental`)  
- CTEs and subqueries  

> - Basic aggregations and joins 

Reviewer: would add a line for semantic layer queries using semantic layer jinja functions

> Limitations?

Reviewer: ```suggestion
```

Reviewer: we can add these later on!

### Example

Let's use an example to illustrate how to run queries in the Query page:

- A Jaffle shop wants to count unique orders and unique customers to understand whether they can expand their Jaffle shop business to other parts of the world.
- To express this logic in SQL, the analyst assigned to this project will:
  - Write a SQL query to calculate the number of unique orders and customers. For example: <br />
    ```sql
    select 
        date_trunc('year', ordered_at) as order_year,
        count(distinct o.customer_id) as unique_customers,
        count(distinct o.order_id) as unique_orders
    from {{ ref('orders') }} o
    join {{ ref('customers') }} c
        on o.customer_id = c.customer_id
    group by 1
    order by 1

>     ```

Reviewer: ```suggestion
    ```sql
    with 

    orders as (
        select * from {{ ref('orders') }}
    ),

    customers as (
        select * from {{ ref('customers') }}
    )
    
    select 
        date_trunc('year', ordered_at) as order_year,
        count(distinct orders.customer_id) as unique_customers,
        count(distinct customers.city) as unique_cities,
        sum(orders.total_amount) as total_order_revenue
    from orders
    join customers
        on orders.customer_id = customers.customer_id
    group by 1
    order by 1
    ```
```

Author: hey @dave-connors-3 , had to change the sql bc i was getting some errors with `customers.city` and `orders.total_amount` since those columns did'nt exist. so changed it to the below:

```sql
    with 

    orders as (
        select * from {{ ref('orders') }}
    ),

    customers as (
        select * from {{ ref('customers') }}
    )

    select 
        date_trunc('year', ordered_at) as order_year,
        count(distinct orders.customer_id) as unique_customers,
        count(distinct orders.location_id) as unique_cities,
        to_char(sum(orders.order_total), '999,999,999.00') as total_order_revenue
    from orders
    join customers
        on orders.customer_id = customers.customer_id
    group by 1
    order by 1
```

The analyst can now run the query by clicking the **Run** button and:
- Explore and [view the results](#view-results) in the **Results** tab.
- [View the details](#view-details) of the query in the **Details** tab.
- [View the chart results](#chart-results) of the query results in the **Chart** tab.
- [View the history](#query-history) of queries and their statuses (like Success, Error, Pending) using the **Query history** icon.
- [Use dbt Explorer](#use-dbt-explorer) to view the lineage and resources of the query.

> - Likewise if they want to promote the query to a model, they [click **Develop in the IDE**](/docs/collaborate/navigate-query-page) on the top right of the Query page to create a new model using the dbt Cloud IDE.

Reviewer: would also want to mention users can develop in visual editor

### View results

Using the same example, the analyst can perform some exploratory data analysis by running the query and:

> - View results: See up to **500 rows** in a table format.

Reviewer: i think we can remove the 500 rows bit. there's no limitation of rows at the moment, we just paginate the results

> - Export results: On the top right of the table, click the the three ellipsis (`...`) button and select **Export to CSV** to export the dataset.

Reviewer: ```suggestion
- Export to CSV: On the top right of the table, click the the three ellipsis (`...`) button and select **Export to CSV** to export the dataset.
```

>   - Note? Any notes here? 

Reviewer: ```suggestion
```

> <Lightbox src="/img/docs/collaborate/query-page/qp-results.jpg" width="95%" title="Query page Results tab" />

Reviewer: will have to update this screenshot since i changed the sql for you hehe

### View details
To view the details of the query, click on the **Details** tab to see:

> - Query metadata: Title, description, supplied SQL, compiled SQL

Reviewer: Title and description are AI generated! This is something we should highlight

> - Connection details: Warehouse, database, schema, role, user

Reviewer: some of those are warehouse specific, so might just say like "pertinent connection information" or something

> - Query details: Query duration, status, columns, rows

Reviewer: ```suggestion
- Query details: Query duration, status, column count, row count
```

<Lightbox src="/img/docs/collaborate/query-page/qp-details.jpg" width="95%" title="Query page Details tab" />

### Chart results

The Query page also supports charting results. An analyst can:
- Select the chart type and columns to visualize.

> - Choose from **line, bar, or dot charts**.

Reviewer: ```suggestion
- Choose from **line chart, bar chart, or scatterplot**.
```

Reviewer: (we will update the menu) also this list will likely grow over time!

> - Charts work best with numerical or categorical data.

Reviewer: ```suggestion
```

<Lightbox src="/img/docs/collaborate/query-page/qp-chart.jpg" width="95%" title="Query page Chart tab" />

### Query history

The Query page also supports query history. The analyst can view the history of queries and their statuses (like Success, Error, Pending), and select a query to re-run and view the results.

| Feature           | Availability |
|------------------|--------------|
| View or search for past queries | ✅ Yes |
| Filter by status  | ✅ Yes (Success, Error, Pending) |
| Export query history | ❌ No |

> How long is history stored?

Reviewer: history is stored indefinitely at the moment!

> | Export query history | ❌ No |

Reviewer: idk that we need this callout here for an unsupported feature until it's requested!

<Lightbox src="/img/docs/collaborate/query-page/qp-query-history.png" width="95%" title="Query page Query history icon" />

> <!-- add copilot section here if available 

Reviewer: we'll have something stood up here very soon! let's leave this out for now

## Query with dbt Copilot

dbt Cloud Copilot is a feature that allows you to query your data with natural language. It is available in the Query page....
-->

## Use dbt Explorer

View dbt Explorer directly in the Query page to access your project lineage and project resources with access to tables, columns, metrics, and dimensions, and more — all integrated in the Query page interface. 

This integrated view allows you to maintain your query workflow, while getting more context on models, semantic models, metrics, macros, and more. The integrated Explorer view provides you with:
- Same search capabilities as Explorer
- Allows you to narrow down displayed objects by type

To access dbt Explorer, click on the **Explorer** icon on the top right of the Query page.

<Lightbox src="/img/docs/collaborate/query-page/qp-explorer.png" width="90%" title="Query page integrated with dbt Explorer" />
 
Questions:

> - Do we need a job run? 

Reviewer: yes, all the rules about metadata visibility in explorer apply here as well. Not sure this nuance needs to live on this docs page

- Are the following available in the private beta?

>   - Automatic highlighting: Selecting an object in the console shows what?

Reviewer: There is a hyperlink from `ref` to the corresponding explorer page, but no highlights

## Considerations 

> - Currently, there’s a limit of 500 returned queries (how to bypass or not possible?). Support for more rows coming soon! 

Reviewer: I don't think this is true at the moment! spiking on work to add a limit UI

Author: ok thanks @dave-connors-3 ! was going by [this](https://linear.app/dbt-labs/issue/SL-3205/load-table-results-from-graphql-api) a couple of months ago

> - Although not currently supported, you will have the ability to save, organize, and share frequently used queries – allowing you to promote consistent use across teams without leaving the Query page.

Reviewer: Currently you can save / bookmark queries for yourself. you can't share them yet, that is coming soon

> - Query page is available in development environment only, with support for multiple environments coming soon! 

Reviewer: The query page _uses your development credentials to query_. You have the ability to query against any object in any environment. 

current: every jinja function uses `defer --favor-state` to resolve jinja
coming soon: ability to select the environment you use to resolve your refs.

> - Any access considerations?

Reviewer: I am honestly not positive -- does this refer to RBAC in dbt Cloud or access to resources in the warehouse or some special third thing?

> - Export to CSV limitations or control?

Reviewer: not at the moment, but likely something we will explore eventually

<!-- this can move to another page -->

## FAQs
- What’s the difference between Query page and dbt Explorer?
	- That’s a great question! Explorer helps you understand your dbt project's structure, resources, lineage, and metrics, offering context for your data
	The Query page builds on that context, allowing you to write, run, and iterate on SQL queries directly in dbt Cloud. It’s designed for ad-hoc or exploratory analysis and empowers business users and analysts to explore data, ask questions, and collaborate seamlessly.
	Explorer provides the context, while Query page enables action.

> - What happens if I exceed the 500-row limit?

Reviewer: remove for now

> - I can’t view my lineage or resources in the integrated dbt Explorer view?

Reviewer: not sure what this is?

    ),

    customers as (
        select * from {{ ref('customers') }}
    )
    
    select 
        date_trunc('year', ordered_at) as order_year,
        count(distinct orders.customer_id) as unique_customers,
        count(distinct customers.city) as unique_cities,
        sum(orders.total_amount) as total_order_revenue
    from orders
    join customers
        on orders.customer_id = customers.customer_id
    group by 1
    order by 1
    ```

The analyst can now run the query by clicking the **Run** button and:
- Explore and [view the results](#view-results) in the **Results** tab.
- [View the details](#view-details) of the query in the **Details** tab.
- [View the chart results](#chart-results) of the query results in the **Chart** tab.
- [View the history](#query-history) of queries and their statuses (like Success, Error, Pending) using the **Query history** icon.
- [Use dbt Explorer](#use-dbt-explorer) to view the lineage and resources of the query.
- Likewise if they want to promote the query to a model, they [click **Develop in the IDE**](/docs/collaborate/navigate-query-page) on the top right of the Query page to create a new model using the dbt Cloud IDE.

> 

Author: will the ui element change so its not just 'develop in ide'?

---
title: "Access the Query page"
description: "Learn how to access the Query page and run queries"
sidebar_label: "Access and run queries"
tags: [Query page]
image: /img/docs/query-page/qp-chart.jpg
---

# Access the Query page interface <Lifecycle status="beta,enterprise" />

<IntroText>
Learn how to access the Query page, run queries, and view results.
</IntroText>

:::tip
Query page is available in private beta to Enterprise accounts. To join, please reach out to your account manager. 
:::

The Query page provides a rich console experience with editor navigation. You can expect the Query page to:
- Enable you to write SQL queries, with the option to open multiple tabs 
- Have SQL + dbt autocomplete suggestions and syntax highlighting
- Bookmark SQL queries
- View the results of the query and its details using the **Results** or **Details** tabs
- Create a visualization of your query results using the **Chart** tab
- View the history of queries and their statuses (like Success, Error, Pending) using the **Query history** icon
- Use dbt Copilot to generate or edit SQL queries using natural language prompts

> - Integrate with [dbt Copilot](/docs/cloud/dbt-copilot), [dbt Explorer](/docs/collaborate/explore-projects), [dbt Cloud IDE](/docs/cloud/dbt-cloud-ide/develop-in-the-cloud), and [Visual Editor](/docs/cloud/visual-editor) to provide a seamless experience for data exploration, AI-assisted writing, and collaboration.

Reviewer 2: ```suggestion
- Integrate with [dbt Copilot](/docs/cloud/dbt-copilot), [dbt Explorer](/docs/collaborate/explore-projects), [dbt Cloud IDE](/docs/cloud/dbt-cloud-ide/develop-in-the-cloud), and [Visual Editor](/docs/cloud/visual-editor) to provide a seamless experience for data exploration, AI-assisted writing, and collaboration
```

### Chart results

Kimiko can visualize the chart results of the query by clicking on the **Chart** tab to:
- Select the chart type using the chart icon.
- Choose from **line chart, bar chart, or scatterplot**.
- Select the axis and columns to visualize using the **Chart settings** icon.

<Lightbox src="/img/docs/query-page/qp-chart.jpg" width="95%" title="Query page Chart tab" />

### Query history

Kimiko can also view the history of queries and their statuses (like Success, Error, Pending) using the **Query history** icon:
- She can select a query to re-run to view the results. 
- She can search for past queries and filter by status.
- For each query, she can click on the ellipsis **(`...`)** button to open the query in a new tab or copy the SQL.

The query history is stored indefinitely.

<Lightbox src="/img/docs/query-page/qp-query-history.png" width="95%" title="Query page Query history icon" />

### Use dbt Explorer

Kimiko accesses [dbt Explorer](/docs/collaborate/explore-projects) directly in the Query page to view the project lineage and project resources with access to tables, columns, metrics, and dimensions, and more — all integrated in the Query page interface. 

> This integrated view allows her and other users maintain their query workflow, while getting more context on models, semantic models, metrics, macros, and more. The integrated Explorer view comes with:

Reviewer 2: ```suggestion
This integrated view allows her and other users to maintain their query workflow, while getting more context on models, semantic models, metrics, macros, and more. The integrated Explorer view comes with:
```

- Same search capabilities as Explorer
- Allows users to narrow down displayed objects by type
- Hyperlink from SQL code `ref` to the corresponding Explorer page

To access dbt Explorer, click on the **Explorer** icon in the [Query console sidebar menu](/docs/collaborate/navigate-query-page#query-console-sidebar-menu).

<Lightbox src="/img/docs/query-page/qp-explorer.png" width="90%" title="Query page integrated with dbt Explorer" />

## Considerations 

> - You can save and bookmark frequently used queries for yourself. Coming soon: Sharing those queries with others.

Reviewer 2: Hiya @mirnawong1 

As a non-blocker - maybe "Coming soon: Sharing those queries with others" could be italicised or added to a new line / new bullet to make it standout (but not overpower the other info)?

Kind Regards
Natalie

- Check the [**Query history**](#query-history) for status and past runs
- Use [**dbt Explorer**](#use-dbt-explorer) to explore model lineage and context

:::tip Want to turn a query into a model?
Coming soon &mdash; you'll be able to access the [dbt Cloud IDE](/docs/cloud/dbt-cloud-ide/develop-in-the-cloud) or [Visual Editor](/docs/cloud/visual-editor) from the [Query console menu](/docs/collaborate/navigate-query-page#query-console-menu) to promote your SQL into a reusable dbt model &mdash; all within dbt Cloud!
:::

### View results

Using the same example, Kimiko can perform some exploratory data analysis by running the query and:

- Viewing results in **Results** tab &mdash; View the paginated results of the query.
- Sorting results &mdash; Click on the column header to sort the results by that column.
- Exporting to CSV &mdash; On the top right of the table, click the three-dot ellipsis (`...`) button and select **Export to CSV** to export the dataset.
<Lightbox src="/img/docs/query-page/qp-export-csv.jpg" width="95%" title="Query page Export to CSV" />

### View details
Kimiko can also view the details of the query by clicking on the **Details** tab:

> - **Query metadata** &mdash; dbt Copilot AI-generated title and description. Along with the supplied SQL and compiled SQL.

Reviewer: ```suggestion
- **Query metadata** &mdash; dbt Copilot AI-generated title and description, the supplied SQL and corresponding compiled SQL.
```
